### PR TITLE
feat: track number of channels joined per user

### PR DESCRIPTION
**Pull request by lauti7**
This PR: 
- calculate the average, max, min, total channel joined, and total created channels by users

### CODE REVIEW

File: src/user.rs [hunk 1]
```diff
@@ -536,7 +583,13 @@ fn pick_random_action(
...
     let mut rng = rand::thread_rng();
     if rng.gen_ratio(probability_to_act as u32, 100) {
-        if rng.gen_ratio(1, 75) {
+        if rng.gen_ratio(1, 3) {
+            SocialAction::SendMessage(RoomType::DirectMessage)
+        } else if channels_enabled && rng.gen_ratio(1, 5) {
+            SocialAction::SendMessage(RoomType::Channel)
+        } else if rng.gen_ratio(1, 3) {
+            SocialAction::AddFriend
+        } else if rng.gen_ratio(1, 75) {
```
Comment (moliva) [inline src/user.rs:592]: Is this change made on purpose? It might change the behavior significantly. Maybe it would be better to take to another PR.

File: src/simulation.rs [hunk 2]
```diff
@@ -141,6 +142,45 @@ impl Simulation {
...
+        let count = ready_users.count();
+        println!("CHANNEL STATS - ready users count: {}", count);
+        println!(
+            "CHANNELS STATS - \n max number user joined: {} \n  min number user joined : {} \n avg joined channel: {} \n total joined: {} \n total created: {}",
+            channel_stats.0,
+            channel_stats.1,
+            (channel_stats.2) as f64 / (count) as f64,
+            channel_stats.2,
+            channel_stats.3.len()
+        );
```
Comment (moliva) [inline src/simulation.rs:182]: Avoid printing everything inline in favor of adding this data to the `final_report`.

File: src/simulation.rs [hunk 3]
```diff
@@ -141,6 +142,45 @@ impl Simulation {
         // wait for report response
         let final_report = events_report.await.expect("events collection to end");
 
+        let ready_users = self.entities.values().filter_map(|entity| {
```
Comment (moliva) [inline src/simulation.rs:145]: Let's extract this to a separate function to avoid having too many details in the `Simulation#run`.

File: src/simulation.rs [hunk 4]
```diff
@@ -141,6 +142,45 @@ impl Simulation {
...
+            (0, usize::MAX, 0, HashSet::new()),
+            |(max, min, total_chans_of_users, mut channels_created), user| {
+                let u = user.try_read();
+
+                if let Ok(u) = u {
+                    log::debug!("getting user {} channel stats", u.localpart);
+                    let results = u.get_user_channels_stats((
+                        max,
+                        min,
+                        total_chans_of_users,
```
Comment (moliva) [inline src/simulation.rs:163]: What's the idea of this number, do we want the number of joins or the number of unique channels joined?

**Review by moliva — COMMENTED** [review]
Leaving some requests and questions.

File: src/simulation.rs (hunk 4 above)
Comment (lauti7) [reply src/simulation.rs:163]: `channels_created` refers to unique channels joined (i should change the name) and `total_chans_of_user` is the sum of each user channel (joined) count to be able to calculate an average.

File: src/user.rs (hunk 1 above)
Comment (lauti7) [reply src/user.rs:592]: Okay, I will move this to another PR. You r right.

File: src/simulation.rs [hunk 5]
```diff
@@ -141,7 +153,78 @@ impl Simulation {
...
+        if context.config.simulation.channels_load {
+            let collect = self.get_channels_info();
+            channels_info = Some(collect);
+        }
+
+        self.store_report(&final_report, channels_info).await;
+    }
+
+    fn get_ready_entities(&self) -> impl Iterator<Item = &Arc<RwLock<User>>> {
+        let ready_users = self.entities.values().filter_map(|entity| {
```
Comment (moliva) [inline src/simulation.rs:167]: nit: Return this expression instead of setting a var.

**Review by moliva — APPROVED** [review]
